### PR TITLE
New UI for insert image feature

### LINKED ISSUE #15149: Better entry point for image file managers and insert image changes
## 📝 Provide a description of the new feature

As we observe with CKBox, when an editor has an integrated file manager, discoverability of uploading images with it is low. Usually, such plugin has an icon that's not that that recognizable as an image one.

CKEditor 4 had a pretty visible option of _Browse server_.

We would move to a bit of different direction, opening up a possibility to add to an image split button more options. One of which to file managers. In the future, such extensibility could be used by other features.

image

### Scope

After doing a PoC #15195, we decided to abandon the route of declarative UI in this area.

We also had a few discussions on the icon set. We decided that we will switch the icon of the `uploadImage` component. The primary objective is to effectively communicate the function of the button. The generic image icon, in other applications, either opens a modal (CKEditor 4, Tiny) or a dropdown (GDocs). In our case, the upload screen opens immediately, therefor a more specific icons was prepared.

#### Features and changes

*   `insertImage` should be intelligent when it comes to registering default actions:
    *   The default action on the split button is the first available on the list below
        *   if there's only one action, there's no dropdown arrow, unless it is insert with link.
    *   The default list in the split button is:
        *   Upload from computer (when `uploadImage` integration is provided)
        *   Insert with file manager (when `ckbox` or `ckfinder` integration is provided)
        *   Insert via URL (insert image via url integration is provided)
*   Remember when selection is on image, the label changes:
    *   Replace from computer
    *   Replace with file manager.
    *   Update image URL.

#### Integrator

*   As an integrator, I can add additional components for the insert image button.
*   As an integrator, I can override the default order on the list.

#### Docs

*   Document insert image component. #15304
*   In documentation, replace everywhere `uploadImage` with `insertImage`. #15371
*   To the demos where `insertImage` was added, add powered by CKBox config for

#### Resources

*   [Figma](https://www.figma.com/file/8pE23Yw147vLEpy4FbUT4M/%5BCKBox%5D-integration-improvements?type=design&node-id=172-530&mode=design&t=khjUWkiTLgfOyoLf-4)

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

### LINKED ISSUE #15195: Create toolbar as split-button PoC
We need a PoC to uncover some details:

*   How to register as split button as a [toolbar grouping](https://ckeditor.com/docs/ckeditor5/latest/features/toolbar/toolbar.html#grouping-toolbar-items-in-drop-downs-nested-toolbars).
*   How to add a main action to the split button via toolbar config.
*   How to define the toolbar grouping as vertical.
*   Can we add a more complex component to a toolbar (like insertViaUrl).

The idea is that we would like to give integrators possibility to define a complex image upload component.

_Originally posted by @Witoso in https://github.com/ckeditor/ckeditor5/issues/15149#issuecomment-1757254114_

### LINKED ISSUE #15303: Improve and finish the PoC of the new `insertImage`
Production-ready setup of the `insertImage` component.
Finishing: #15255

### PR DESCRIPTION
**Pull request by niegowski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Feature (image): Introduced the image insert dropdown as a consistent UI to insert images through different available integrations like image upload, insert image with asset manager, and insert image via URL. Closes #15303. Closes #15149.

Other (image): The `ImageUploadUI` plugin is loaded by default when the `ImageBlock` or `ImageInline` plugins are loaded. See #15149.

Other (list, ui): The `CollapsibleView` moved from the `list` package to the `ui` package. See #15149.

Other (ui): The `SplitButtonView` constructor and `createDropdown()` helper accepts an instance of a `ButtonView` as an action view customization. See #15149.

Other (upload): The `FileDialogButtonView` is now an instance of the `ButtonView`, not just a wrapper on it. See #15149.

Internal (ckbox, ckfinder, image): The `ImageUploadUI`, `CKBoxUI`, `CKFinderUI`, and `ImageInsertViaUrlUI` plugins are registering integrations in the `ImageInsertUI`.

Internal (core): Added icons for `ImageUploadUI` and its integrations. Added translation contexts for common labels for images insert/replace via file manager.

MINOR BREAKING CHANGE (list): The `CollapsibleView` moved from the `list` package to the `ui` package. You can import it like this: `import { CollapsibleView } from '@ckeditor/ckeditor5-ui';`

---

### CODE REVIEW

**Review by arkflpc — APPROVED** [review]

**Review by arkflpc — APPROVED** [review]

File: packages/ckeditor5-image/src/imageinsert/ui/imageinsertformview.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+	/**
+	 * Focuses the first {@link #_focusables focusable} in the form.
+	 */
+	public focus(): void {
+		this._focusCycler.focusFirst();
+	}
+}
+
+function isViewWithFocusCycler( view: View ): view is View & { focusCycler: FocusCycler } {
```
Comment (pszczesniak) [inline packages/ckeditor5-image/src/imageinsert/ui/imageinsertformview.ts:167]: Missing description.

File: packages/ckeditor5-theme-lark/theme/ckeditor5-ui/components/collapsible/collapsible.css [hunk 2]
```diff
@@ -32,7 +32,7 @@
 	}
 
 	& > .ck-collapsible__children {
-		padding: 0 var(--ck-spacing-large) var(--ck-spacing-large);
+		padding: var(--ck-spacing-medium) var(--ck-spacing-large) var(--ck-spacing-large);
```
Comment (pszczesniak) [inline packages/ckeditor5-theme-lark/theme/ckeditor5-ui/components/collapsible/collapsible.css:35]: Why top padding is changed from `large` into `medium`? When button is focused it looks odd when label in on the box-shadow outline:  
 
Screenshot 2023-12-05 at 15 44 30

But as i understand decision was made - i just wanted to show my doubts.

**Review by pszczesniak — APPROVED** [review]
LGTM :+1: 

One missing description - but it's not a blocker.

**Comment by pszczesniak** [comment]
### One side comment from me:

I was/am a bit confused when two icons/buttons with same functionality are labeled differently:





Don't know why it is done this way.

File: packages/ckeditor5-theme-lark/theme/ckeditor5-ui/components/collapsible/collapsible.css (hunk 2 above)
Comment (niegowski) [reply packages/ckeditor5-theme-lark/theme/ckeditor5-ui/components/collapsible/collapsible.css:35]: It's increased from `0` to `medium` 😀

Comment (pszczesniak) [reply packages/ckeditor5-theme-lark/theme/ckeditor5-ui/components/collapsible/collapsible.css:35]: Missed the `0` , sorry! but still, large looks a bit better IMO ;)

Screenshot 2023-12-05 at 16 17 17

**Comment by niegowski** [comment]
> Don't know why it is done this way.

To give context. In the dropdown, the user already has a context that it's a dropdown to insert an image. The main button does not have the context, so an extended label is used to make it clear that it's inserting images, not any file.

File: packages/ckeditor5-theme-lark/theme/ckeditor5-ui/components/collapsible/collapsible.css (hunk 2 above)
Comment (niegowski) [reply packages/ckeditor5-theme-lark/theme/ckeditor5-ui/components/collapsible/collapsible.css:35]: When it's larger it looks strange when the button outline is not visible.

**Pull request merged**